src/overlays/actors/ovl_En_Rr/z_en_rr.c

```diff
@@ -21,51 +83,775 @@ const ActorInit En_Rr_InitVars = {
+                Item_DropCollectible(globalCtx, &dropPos, ITEM00_RUPEE_PURPLE);
+                break;
+            case RR_DROP_RED_RUPEE:
+                Item_DropCollectible(globalCtx, &dropPos, ITEM00_RUPEE_RED);
+                break;
+            case RR_DROP_RANDOM_RUPEE:
+            default:
+                Item_DropCollectibleRandom(globalCtx, &this->actor, &dropPos, 12 << 4);
+                break;
+        }
+        Actor_Kill(&this->actor);
+    } else if (this->frameCount == 88) {
+        Vec3f pos;
+        Vec3f vel;
+        Vec3f accel;
+
+        pos.x = this->actor.posRot.pos.x;
+        pos.y = this->actor.posRot.pos.y + 20.0f;
+        pos.z = this->actor.posRot.pos.z;
+        vel.x = 0.0f;
+        vel.y = 0.0f;
+        vel.z = 0.0f;
+        accel.x = 0.0f;
+        accel.y = 0.0f;
+        accel.z = 0.0f;
+
+        EffectSsDeadDb_Spawn(globalCtx, &pos, &vel, &accel, 100, 0, 255, 255, 255, 255, 255, 0, 0, 1, 9, 1);
+    } else {
+        Math_SmoothScaleMaxF(&this->actor.scale.x, 0.0f, 1.0f, this->shrinkRate);
+        Math_SmoothScaleMaxF(&this->shrinkRate, 0.001f, 1.0f, 0.00001f);
+        this->actor.scale.z = this->actor.scale.x;
+    }
+}
+
+void EnRr_Retreat(EnRr* this, GlobalContext* globalCtx) {
+    if (this->actionTimer == 0) {
+        this->retreat = false;
+        this->actionFunc = EnRr_Approach;
+    } else {
+        Math_SmoothScaleMaxMinS(&this->actor.shape.rot.y, this->actor.yawTowardsLink + 0x8000, 0xA, 0x3E8, 0);
+        this->actor.posRot.rot.y = this->actor.shape.rot.y;
+        if (this->actor.speedXZ == 0.0f) {
+            EnRr_SetSpeed(this, 2.0f);
+        }
+    }
+}
+
+void EnRr_Stunned(EnRr* this, GlobalContext* globalCtx) {
+    if (this->actor.dmgEffectTimer == 0) {
+        this->stopScroll = false;
+        if (this->hasPlayer) {
+            EnRr_SetupReleasePlayer(this, globalCtx);
+        } else if (this->actor.colChkInfo.health != 0) {
+            this->actionFunc = EnRr_Approach;
+        } else {
+            EnRr_SetupDeath(this);
+        }
+    }
+}
+
+void EnRr_Update(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    EnRr* this = THIS;
+    s32 i;
+
+    this->frameCount++;
+    if (!this->stopScroll) {
+        this->scrollTimer++;
+    }
+    if (this->actionTimer != 0) {
+        this->actionTimer--;
+    }
+    if (this->grabTimer != 0) {
+        this->grabTimer--;
+    }
+    if (this->ocTimer != 0) {
+        this->ocTimer--;
+    }
+    if (this->invincibilityTimer != 0) {
+        this->invincibilityTimer--;
+    }
+    if (this->effectTimer != 0) {
+        this->effectTimer--;
+    }
+
+    Actor_SetHeight(&this->actor, 30.0f);
+    EnRr_UpdateBodySegments(this, globalCtx);
+    if (!this->isDead && ((this->actor.dmgEffectTimer == 0) || !(this->actor.dmgEffectParams & 0x4000))) {
+        EnRr_CollisionCheck(this, globalCtx);
+    }
+
+    this->actionFunc(this, globalCtx);
+    if (this->hasPlayer == 0x3F80) { // checks if hasPlayer is 1.0f
```

Comment:
What is this exactly? Some kind of truncated float?

Reply:
I think it's checking for if `1.0f` has been mistakenly written to that address instead of `1`. I'm not sure how such a thing could happen, but `hasPlayer` is definitely a boolean.

Reply:
Interesting. Odd that they used an s16. Is it possible to match by doing something like `if (this->hasPlayer == 1.0f)`? Also, maybe put a comment explaining what you just explained to me, like, `// @bug The programmer who worked on this accidentally used a floating point 1.0 instead of a 1 here`, making it clear it's a bug.

Reply:
I don't think it's a bug. Given that if the check succeeds it asserts and crashes the game, I think checking if `1.0f` has been written there instead of 1 is an intentional error check.

Reply:
Gotcha.

Reply:
1.0f is 0x3F800000, not 0x3F80. So this comment is technically incorrect

Reply:
Now technically correct. (The best kind of correct?)